dev/sql_auditing.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+#include <fstream>
+#include "error_code.h"
+#include <chrono>
+#include <ctime>  // For std::localtime and std::tm
+#include <iomanip>  // For std::put_time
+#include <string>
+
+enum class auditing_behavior : signed char { OFF = 0, ON = 1 };
+
+class sql_auditor_settings;
+
+class sql_auditor {
+
+    std::ofstream log_file;
+    friend class sql_auditor_settings;
+    sql_auditor();
+    void open();
+    inline static sql_auditor& auditor() {
+        static sql_auditor auditor{};
+        return auditor;
+    }
+
+  public:
+    static void log(const std::string& message);
+};
+
+class sql_auditor_settings {
+    std::string destination_file = "sql_auditor.txt";
+    friend class sql_auditor;
+    sql_auditor_settings() {}
+    inline static sql_auditor_settings& settings() {
```

Comment:
the destination file could be part of the storage object but then we could not log the initial instructions (i.e. make_storage) -- what if we want to see the results of make_storage()?

Reply:
since the call to sql_auditor::log() occurs in util.h, how do we know which storage it belongs to? I am talking about e.g. void perform_exec()??? we only know the sqlite3* but that does not map to storage so its no use

Reply:
we could intercept on_open to register every connection used by a storage_base?

Reply:
and create a map of sqlite* pointers and storage_base* pointers!!